src/hooks/useTracking.js

```diff
@@ -32,103 +28,103 @@ const getUserId = () => {
-        }
-      } else {
-        pausedTime = Date.now()
-      }
-    }
-    document.addEventListener('visibilitychange', onVisibilityChanged)
+  let start
+  let pageviewToken
+  let msPaused
+  let pausedTime
+
+  const initEvent = () => {
+    start = Date.now()
+    pageviewToken = generateId()
+    msPaused = 0
+    pausedTime = null
+  }
 
-    const sendBeacon = () => {
+  const onShow = () => {
+    const [, changed] = getSessionId()
+    if (changed) {
+      // The page has been hidden for too long, the session has expired.
+      // We consider this a new pageview, reset the event data (start time, etc)
+      initEvent()
+    } else if (pausedTime) {
       const now = Date.now()
-      const totalTime = now - start
+      msPaused += now - pausedTime
+      pausedTime = null
+    }
+  }
 
-      const event = {
-        event: {
-          /* eslint-disable camelcase */
-          user_id: userId,
-          session_id: getSessionId(),
-          pageview_token: pageviewToken,
-          client_type: 'kaios-app',
-          app_version: appVersion(),
-          referring_domain: 'kaios-app',
-          load_dt: new Date(start).toISOString(),
-          page_open_time: Math.round(totalTime),
-          page_visible_time: Math.round(totalTime - msPaused),
-          section_count: sectionCount,
-          opened_section_count: trackingRef.openedSectionCount,
-          page_namespace: namespace,
-          is_main_page: isSearch,
-          is_search_page: isSearch
-          /* eslint-enable camelcase */
-        },
-        revision: SCHEMA_REV,
-        schema: SCHEMA_NAME,
-        webHost: `${language}.wikipedia.org`,
-        wiki: `${language}wiki`
-      }
+  const onHide = () => {
+    pausedTime = Date.now()
+    // Log the event now since we don't know if the user is ever coming back.
+    logInukaPageView()
+  }
 
-      sendEvent(event)
+  const onVisibilityChanged = () => {
+    document.visibilityState === 'visible' ? onShow() : onHide()
+  }
 
-      document.removeEventListener('visibilitychange', onVisibilityChanged)
-      window.removeEventListener('pagehide', sendBeacon)
+  const logInukaPageView = () => {
+    const now = Date.now()
+    const totalTime = now - start
+    const [sessionId] = getSessionId()
+
+    const event = {
+      /* eslint-disable camelcase */
+      user_id: userId,
+      session_id: sessionId,
+      pageview_token: pageviewToken,
+      client_type: 'kaios-app',
+      app_version: appVersion(),
+      referring_domain: 'kaios-app',
+      load_dt: new Date(start).toISOString(),
+      page_open_time: Math.round(totalTime),
+      page_visible_time: Math.round(totalTime - msPaused),
+      section_count: sectionCount,
+      opened_section_count: trackingRef.openedSectionCount,
+      page_namespace: namespace,
+      is_main_page: isSearch,
+      is_search_page: isSearch
+      /* eslint-enable camelcase */
     }
-    window.addEventListener('pagehide', sendBeacon)
```

Comment:
pagehide event not useful here?

Reply:
when detect complete exit the app.

Reply:
How do you test it on the device?

Reply:
Previously, I was testing it on WebIDE

and add a special local storage name in the pagehide event, after the app exit, launch it again and see whether this special name exists.

Reply:
I couldn't find a case where visibilitychanged was not enough. Please let me know if you find one.

Reply:
Ok, I saw the comment 

```
// Log the event now since we don't know if the user is ever coming back.
```

So it makes sense to send the event when visibility is hidden 👍 no need pagehide then